runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -10393,227 +10393,637 @@ static TR::Register *inlineIsAssignableFrom(TR::Node *node, TR::CodeGenerator *c
-    return hashReg;
-}
+static TR::Register*
+hashCodeHelper(TR::Node *node, TR::CodeGenerator *cg, TR::DataType elementType,
+               TR::Node *hashNode, bool isSigned)
+   {
+   TR::Compilation *comp = cg->comp();
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(comp->fe());
+   bool isLE = comp->target().cpu.isLittleEndian();
+   const bool nonZeroInitial = hashNode != NULL && !hashNode->isConstZeroValue();
+
+
+   TR::Node *valueNode = node->getFirstChild();
+   TR::Node *offsetNode = node->getSecondChild();
+   TR::Node *countNode = node->getThirdChild();
+
+   TR::Register *valueReg = cg->gprClobberEvaluate(valueNode);
+   TR::Register *endReg = cg->gprClobberEvaluate(offsetNode);
+   TR::Register *vendReg = cg->gprClobberEvaluate(countNode);
+   TR::Register *tempReg = cg->allocateRegister();
+   TR::Register *constant0Reg = cg->allocateRegister();
+   TR::Register *multiplierAddrReg = cg->allocateRegister();
+   TR::Register *condReg = cg->allocateRegister(TR_CCR);
+
+   TR::Register *multiplierReg = cg->allocateRegister(TR_VRF);
+   TR::Register *vtmp1Reg = cg->allocateRegister(TR_VRF);
+   TR::Register *vtmp2Reg = cg->allocateRegister(TR_VRF);
+   TR::Register *vconstant0Reg = cg->allocateRegister(TR_VRF);
+   TR::Register *vconstantNegReg = cg->allocateRegister(TR_VRF);
+   TR::Register *vunpackMaskReg = cg->allocateRegister(TR_VRF);
+
+   // we can load 16 bytes at once, but each register can only accumulate 4 values
+   // so for smaller datatypes we need more than one registers
+   TR::Register *low4Reg = cg->allocateRegister(TR_VRF);
+   TR::Register *high4Reg, *third4Reg, *fourth4Reg;
+   TR::Register *vtmp3Reg;
+   switch (elementType)
+      {
+      case TR::Int8:
+         vtmp3Reg = cg->allocateRegister(TR_VRF); // a third register for unpacking
+         third4Reg = cg->allocateRegister(TR_VRF);
+         fourth4Reg = cg->allocateRegister(TR_VRF);
+         // fall through
+      case TR::Int16:
+         high4Reg = cg->allocateRegister(TR_VRF);
+         // fall through
+      case TR::Int32:
+         break;
+      default:
+         TR_ASSERT_FATAL(false, "Unsupported hashCodeHelper elementType");
+      }
+
+   // the hashReg is zero initially, unless specified otherwise
+   TR::Register *hashReg = NULL;
+   if (nonZeroInitial)
+      {
+      hashReg = cg->gprClobberEvaluate(hashNode);
+      }
+   else
+      {
+      hashReg = cg->allocateRegister();
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::XOR, node, hashReg, hashReg, hashReg);
+      }
+
+   TR::LabelSymbol *serialLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *VSXLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *POSTVSXLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *endLabel = generateLabelSymbol(cg);
+
+   // Skip header of the array
+   intptr_t hdrSize = TR::Compiler->om.contiguousArrayHeaderSizeInBytes();
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, valueReg, valueReg, hdrSize);
```

Comment:
Seems to me that adding a load of dataAddr pointer instead of adding headerSize here is only thing holding enabling this for OffHeap, @zl-wang. Not sure if worth doing it separately here or with other platforms later.

Reply:
Given that we're in the final stages of enabling OffHeap I think deferring this for later would be better, to not possibly introduce new issues regarding OffHeap. I added it to the OffHeap TODO list.